google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +384,36 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
-      pendingReceipts.add(ackRequestData);
-    }
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        ApiFutureCallback<AckResponse> callback = new ApiFutureCallback<AckResponse>() {
+          @Override
+          public void onFailure(Throwable throwable) {
+            logger.log(
+                Level.WARNING,
```

Comment:
We also want to drop the message in the failure scenario right? As written, this corresponding message will still be tracked by pendingMessages, so we will continue extending deadlines for this message, which I am assuming we don't want to do.

We could consider calling forget() for this message, so it will be removed from pendingMessages (but this also frees the flow controller, so some additional changes will be required). We could also consider not adding a message to pendingMessages until we are ready to deliver it to the end user (but we would need to confirm this wouldn't cause other unwanted side effects)

Let's look further into these options to determine the best approach forward?